runtime/near-vm-runner/src/wasmer2_runner.rs

```diff
@@ -301,62 +271,245 @@ pub(crate) fn wasmer2_vm_hash() -> u64 {
+        method_name: &str,
+    ) -> Result<(), VMError> {
+        let _span = tracing::debug_span!(target: "vm", "run_method").entered();
+
+        // FastGasCounter in Nearcore and Wasmer must match in layout.
+        assert_eq!(size_of::<FastGasCounter>(), size_of::<wasmer_types::FastGasCounter>());
+        assert_eq!(
+            offset_of!(FastGasCounter, burnt_gas),
+            offset_of!(wasmer_types::FastGasCounter, burnt_gas)
+        );
+        assert_eq!(
+            offset_of!(FastGasCounter, gas_limit),
+            offset_of!(wasmer_types::FastGasCounter, gas_limit)
+        );
+        assert_eq!(
+            offset_of!(FastGasCounter, opcode_cost),
+            offset_of!(wasmer_types::FastGasCounter, opcode_cost)
+        );
+        let gas = import.vmlogic.gas_counter_pointer() as *mut wasmer_types::FastGasCounter;
+
+        let entrypoint = get_entrypoint_index(&*artifact.artifact, method_name)?;
+        unsafe {
+            let instance = {
+                let _span = tracing::debug_span!(target: "vm", "run_method/instantiate").entered();
+                // SAFETY/FIXME: this transmute shouldn't be necessary, but we are limited by
+                // wasmer's API right now.
+                //
+                // An important caveat is that the `'static` lifetime here refers to the lifetime
+                // of `VMLogic` reference to which is retained by the `InstanceHandle` we create.
+                // However this `InstanceHandle` only lives during the execution of this body, so
+                // we can be sure that `VMLogic` remains live and valid at any time.
+                let import =
+                    std::mem::transmute::<_, &mut Wasmer2Imports<'static, 'static>>(&mut import);
+                // SAFETY: we ensure that the tables are valid during the lifetime of this instance
+                // by retaining an instance to `UniversalEngine` which holds the allocations.
+                let handle = artifact
+                    .artifact
+                    .instantiate(
+                        &self,
+                        import,
+                        Box::new(()),
+                        // SAFETY: We have verified that the `FastGasCounter` layout matches the
+                        // expected layout. `gas` remains dereferenceable throughout this function
+                        // by the virtue of it being contained within `import` which lives for the
+                        // entirety of this function.
+                        InstanceConfig::new_with_counter(gas),
+                    )
+                    .map_err(|err| translate_instantiation_error(err, import.vmlogic))?;
+                // SAFETY: being called immediately after instantiation.
+                artifact
+                    .artifact
+                    .finish_instantiation(self, &handle)
+                    .map_err(|err| translate_instantiation_error(err, import.vmlogic))?;
```

Comment:
This begets a question -- do we actually have a test which checks what happens if gas is exhausted when `_start` is called? I don't recall one.

Reply:
We do have one here: https://github.com/near/nearcore/blob/2a033f42eeefea2fbc867e26ec4281132b6ae447/runtime/near-vm-runner/src/tests/runtime_errors.rs#L32-L44